pkg/endpoint/policy.go

```diff
@@ -40,29 +41,89 @@ import (
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/policy"
 	"github.com/cilium/cilium/pkg/revert"
+	"github.com/cilium/cilium/pkg/u8proto"
 
 	"github.com/sirupsen/logrus"
 )
 
-// ProxyID returns a unique string to identify a proxy mapping.
-func (e *Endpoint) ProxyID(npMap policy.NamedPortsMap, l4 *policy.L4Filter) (string, error) {
-	return policy.ProxyIDFromFilter(e.ID, npMap, l4)
+// GetNamedPort returns the port for the given name.
+// Must be called with e.mutex NOT held
+func (e *Endpoint) GetNamedPort(ingress bool, name string, proto uint8) uint16 {
+	if ingress {
+		// Ingress only needs the ports of the POD itself
+		k8sPorts, err := e.GetK8sPorts()
+		if err != nil {
+			if e.logLimiter.Allow() {
+				e.getLogger().WithFields(logrus.Fields{
+					logfields.PortName:         name,
+					logfields.Protocol:         u8proto.U8proto(proto).String(),
+					logfields.TrafficDirection: "ingress",
+				}).WithError(err).Warning("Skipping named port")
+			}
+			return 0
+		}
+		return e.getNamedPortIngress(k8sPorts, name, proto)
+	}
+	// egress needs named ports of all the pods
+	return e.getNamedPortEgress(ipcache.IPIdentityCache.GetNamedPorts(), name, proto)
+}
+
+// GetNamedPortLocked returns port for the given name. May return an invalid (0) port
+// Must be called with e.mutex held.
+func (e *Endpoint) GetNamedPortLocked(ingress bool, name string, proto uint8) uint16 {
+	if ingress {
+		// Ingress only needs the ports of the POD itself
+		return e.getNamedPortIngress(e.k8sPorts, name, proto)
+	}
+	// egress needs named ports of all the pods
+	return e.getNamedPortEgress(ipcache.IPIdentityCache.GetNamedPorts(), name, proto)
+}
+
+func (e *Endpoint) getNamedPortIngress(npMap policy.NamedPortMap, name string, proto uint8) uint16 {
+	port, err := npMap.GetNamedPort(name, proto)
+	if err != nil && e.logLimiter.Allow() {
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.PortName:         name,
+			logfields.Protocol:         u8proto.U8proto(proto).String(),
+			logfields.TrafficDirection: "ingress",
+		}).WithError(err).Warning("Skipping named port")
```

Comment:
Minor followup: This error condition is separate from the exact same message above on line 61 right? It would be helpful to have a way to distinguish the two in case this message is ever logged, so we know which case was hit.

Reply:
@jrajahalme will take care of this as a follow up!